### PR TITLE
Redesign monitoring internals

### PR DESCRIPTION
**Pull request by davetron5000**
Treating the rake task as the public interface here, but this would break anyone using the stuff in lib/monitoring
    
* Created a real data type for the results of checks
* Removed the "per queue" notifier, which was tightly coupled to its checker.  Now librato notifier can handle all of them
* This should make it easier to add more advanced checks/notifications.
    
To get confidence this won't break anything, I basically made the internal changes and kept running the monitoring integration test.  Once I'd made all the changes without that breaking, I then updated the unit tests.
    
(The change in the monitoring integration test was just to add failure messages to the asserts. No actual changes were made)

This is in advance of another change I'm going to make to add stats around the _class_ of failed jobs, so we can treat certain failures as more urgent than others without having to create a custom resque error handler.  Wanted to get feedback on this first, which will make the overall change easier to understand (hopefully :)

### CODE REVIEW

File: lib/monitoring/check_result.rb [hunk 1]
```diff
@@ -0,0 +1,30 @@
+module Monitoring
+  class CheckResult
```
Comment (davetron5000) [inline lib/monitoring/check_result.rb]: This now captures the contract between checker and notifier.  Before it was basically whatever and so it resulted in tight coupling between some checkers and notifiers.

File: lib/monitoring/librato_notifier.rb [hunk 2]
```diff
@@ -11,9 +10,10 @@ def initialize(prefix: nil, logger: Rails.logger, type: :count, unit: "")
     #
     # results:: a hash where the keys represent the source (the resque instance name) and the values
     #           are lists of items to be counted.  The items won't be examined, just counted and used in the metric
-    def notify!(results)
-      results.each do |resque_name,items|
-        log_to_librato(resque_name,@type,@prefix,items.size)
+    def notify!(check_results)
+      check_results.each do |check_result|
+        source = [check_result.resque_name,check_result.scope].compact.join(".")
```
Comment (davetron5000) [inline lib/monitoring/librato_notifier.rb]: This is how the `scope` is used.  Bascially if we think of the resque name as the source, the scope is a refinement inside that resque of what the source is.  In the case of the queue size check, it uses the name of the queue as the scope (so, in theory, we could alert on different queue sizes differently, which is what i intend to do with the failed job classes)

File: lib/monitoring/per_queue_librato_notifier.rb [hunk 3]
```diff
@@ -1,16 +0,0 @@
-module Monitoring
```
Comment (davetron5000) [inline lib/monitoring/per_queue_librato_notifier.rb]: No longer needed given the stronger contract implied by `CheckResult`

File: lib/tasks/monitor.rake [hunk 4]
```diff
@@ -11,7 +11,7 @@ namespace :monitor do
   task :stale_workers => :environment do
     monitor = Monitoring::Monitor.new(
        checker: Monitoring::StaleWorkerCheck.new,
-      notifier: Monitoring::LibratoNotifier.new(prefix: "resque.stale_workers", type: :measure, unit: "workers"))
+      notifier: Monitoring::LibratoNotifier.new(type: :measure, unit: "workers"))
```
Comment (davetron5000) [inline lib/tasks/monitor.rake]: `prefix` was no longer needed as the `CheckResult` can control that

File: lib/monitoring/check_result.rb [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+    attr_reader :resque_name, # Name of the resque the check checked
+                :scope,       # Refined scope inside this resque, if relevenat (e.g. queue name or class name)
```
Comment (geoff2k) [inline lib/monitoring/check_result.rb]: s/relevenat/relevant/

File: lib/monitoring/monitor.rb [hunk 6]
```diff
@@ -1,6 +1,7 @@
 module Monitoring
   class Monitor
-    def initialize(checker: checker,notifier: notifier)
+    def initialize(checker: nil,notifier: nil)
+      raise "both checker and notifier are required" if checker.nil? || notifier.nil?
```
Comment (geoff2k) [inline lib/monitoring/monitor.rb]: In CheckResult we raise a specific ArgumentException if there's a problem with the initialize() parameters -- is there a reason to use RuntimeError here?

Comment (davetron5000) [inline lib/monitoring/monitor.rb]: Nope, just laziness.  Will update

**Pull request merged**